[MrssFormat.php] make sure feed is w3c compliant

madelynseal opened the pull request:
xml attributes need to have certain characters escaped to be valid
the title attribute in this case sometimes has double quotes in it

Initial pull request was #798 but that was a dirty fix and only for the Youtube bridge, this one is for all Mrss feeds

madelynseal committed:
[MrssFormat.php] make sure feed is w3c compliant
xml attributes need to have certain characters escaped to be valid
the title attribute in this case sometimes has double quotes in it

logmanoriginal commented:
I'm a bit confused to what the difference is between before vs. after.

`$title` is already beeing escaped

https://github.com/RSS-Bridge/rss-bridge/blob/458b826871ca7b083589a6d3e840e1fd7116f463/formats/MrssFormat.php#L16

by a function implementing `htmlspecialchars`

https://github.com/RSS-Bridge/rss-bridge/blob/458b826871ca7b083589a6d3e840e1fd7116f463/formats/MrssFormat.php#L115-L117

The only difference is in the additional parameter `ENT_XML1`, which makes no difference in regards to double quotes according to http://php.net/manual/en/function.htmlspecialchars.php

![image](https://user-images.githubusercontent.com/5776685/45318981-d3c06000-b53e-11e8-9df2-31289e999bd3.png)

You can see that double quotes are always escaped, except if `ENT_NOQUOTES` has been set, which is clearly not the case. 

Am I missing something?
Can you share a [Gist](https://gist.github.com/) of a feed that fails for you?

madelynseal commented:
Try [https://www.youtube.com/channel/UCtNIoi7VuApbYOPx-NjcVAA](https://www.youtube.com/channel/UCtNIoi7VuApbYOPx-NjcVAA) and you will see the issue. 

There is a function which "fixes" the title of the youtube channel so instead of having `&quote;` it has actual double quotes. This makes the outputted feed non w3c compliant with channel names that include double quotes.

I hope this explains this issue

teromene commented:
So I went to look for the specifications, and it seems that the elements in the title should *not* be escaped (see http://www.rssboard.org/rss-profile#data-types-characterdata).
The only things that should are `&`, `>` and `<`.

madelynseal commented:
This fix only affects the title of the favicon link which isn't even needed from my knowledge

logmanoriginal commented:
>Try https://www.youtube.com/channel/UCtNIoi7VuApbYOPx-NjcVAA and you will see the issue.

Thanks, that was very helpful :+1: 
I see your issue, basically the feed appears empty like this:

![image](https://user-images.githubusercontent.com/5776685/45503693-e8dbff80-b787-11e8-81ea-1cc5a149c96c.png)

Even though the feed contains a lot of contents: https://gist.github.com/LogMANOriginal/e7aa14116e4e6aaff6e84ca36290d2b8

***

The issue lies in these lines (attribute closed prematurely):

```XML
<image url="https://www.youtube.com/favicon.ico" title="Stanley "Dirt Monkey" Genadek - YouTube" link="https://www.youtube.com/"/>
```

I took another dive into the specification of [`htmlspecialchars`](http://php.net/manual/en/function.htmlspecialchars.php) and it looks like I misunderstood some of the finer details. Long story short, `ENT_XML1` only ensures that a given string contains valid characters for [XML1](https://www.w3.org/TR/2006/REC-xml11-20060816) documents. Double quotes are in fact valid characters (used for attributes).

So, why is this an issue?

The YoutubeBridge returns a title with valid UTF8 encoded text: 

`Stanley "Dirt Monkey" Genadek - YouTube`

The Mrss format ensures the title is valid XML (which it is) and subsequently builds the feed. 

@teromene The title you refer to is the `<title />` tag. You are right, the contents of that element should not be encoded.

Unfortunately the title is also used for the _title_ __attribute__ of the channels' image:

https://github.com/RSS-Bridge/rss-bridge/blob/911bcfb2463f50c28f40ca70e65fa380732c2e02/formats/MrssFormat.php#L93

This is where the attribute was prematurely closed earlier:

```XML
<image url="https://www.youtube.com/favicon.ico" title="Stanley "Dirt Monkey" Genadek - YouTube" link="https://www.youtube.com/"/>
```

So we have to make sure to encode double quotes in (title) attributes. There are two solutions for this particular attribute:

1) Remove the title attribute (bad solution)
2) Properly encode the title for use in attributes (good solution)

@matthewseal Your solution is about right, but there are a few things that need changing to make the intent clear (see my review comments in a minute...)

logmanoriginal commented on formats/MrssFormat.php:
```diff
@@ -78,7 +78,9 @@ public function stringify(){
-
+		
+		/* xml attributes need to have certain characters escaped to be w3c compliant */
+		$titleSafe = htmlspecialchars($title);
```
The variable name should reflect its use, like `imageTitle`. Additionally, make sure to only encode double quotes, so other items don't get encoded as well (like "<" turning into "&lt"). You can use the [`ENT_COMPAT`](http://php.net/manual/en/function.htmlspecialchars.php) flag for that:

```PHP
$imageTitle = htmlspecialchars($title, ENT_COMPAT);
```

>Will convert double-quotes and leave single-quotes alone.
>
>-- http://php.net/manual/en/function.htmlspecialchars.php

logmanoriginal replied:
Travis currently [complains](https://travis-ci.org/RSS-Bridge/rss-bridge/jobs/426200751) about "Whitespace found at end of line", please make sure to remove whitespace from the end of lines.

madelynseal replied:
fixed

logmanoriginal reviewed (changes requested):
Find below my comments

madelynseal committed:
fixes @LogMANOriginal suggested
remove trailing white space
rename $titleSafe to $imageTitle
add flag ENT_COMPAT to htmlspecialchars

logmanoriginal commented:
Merged. Thanks for your patience!

madelynseal commented:
No problem, just glad it is fixed now
